File: mediscan/backend/app/api/community.py
"""
Community API — /api/community/*
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, desc
from pydantic import BaseModel, Field
from typing import Optional, List

from app.database import get_db
from app.models.user import User
from app.models.community import CommunityPost
from app.api.auth import get_current_user
from app.security import InMemoryRateLimiter, validate_barcode

router = APIRouter(prefix="/api/community", tags=["community"])
community_write_limiter = InMemoryRateLimiter(limit=20, window_seconds=60)
vote_limiter = InMemoryRateLimiter(limit=60, window_seconds=60)

VALID_CATEGORIES = ["general", "recipe", "symptom"]


class PostCreate(BaseModel):
    category: str = "general"
    title: str = Field(min_length=5, max_length=160)
    body: str = Field(min_length=1, max_length=5000)
    tags: Optional[List[str]] = Field(default_factory=list, max_length=10)
    product_barcode: Optional[str] = Field(default=None, max_length=50)


class VoteRequest(BaseModel):
    direction: str = Field(pattern="^(up|down)$")


@router.get("/posts")
async def list_posts(
    category: Optional[str] = Query(None),
    page: int = Query(1, ge=1, le=1000),
    per_page: int = Query(15, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    q = select(CommunityPost).order_by(desc(CommunityPost.created_at))
    if category and category in VALID_CATEGORIES:
        q = q.where(CommunityPost.category == category)
    q = q.offset((page - 1) * per_page).limit(per_page)
    result = await db.execute(q)
    posts = result.scalars().all()
    return {
        "posts": [
            {
                "id": p.id,
                "author": p.author_name or "Anonymous",
                "category": p.category,
                "title": p.title,
                "body": p.body[:200] + ("..." if len(p.body) > 200 else ""),
                "tags": p.tags,
                "upvotes": p.upvotes,
                "downvotes": p.downvotes,
                "net_votes": p.upvotes - p.downvotes,
                "product_barcode": p.product_barcode,
                "created_at": p.created_at.isoformat(),
            }
            for p in posts
        ]
    }


@router.post("/posts", status_code=201)
async def create_post(
    body: PostCreate,
    _: None = Depends(community_write_limiter),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    if body.category not in VALID_CATEGORIES:
        raise HTTPException(status_code=422, detail=f"category must be one of {VALID_CATEGORIES}")
    if len(body.title) < 5:
        raise HTTPException(status_code=422, detail="Title too short")
    tags = [(tag or "").strip()[:40] for tag in (body.tags or [])]
    tags = [tag for tag in tags if tag][:10]
    product_barcode = validate_barcode(body.product_barcode) if body.product_barcode else None

    post = CommunityPost(
        user_id=current_user.id,
        author_name=current_user.full_name or current_user.email.split("@")[0],
        category=body.category,
        title=body.title.strip(),
        body=body.body.strip(),
        tags=tags,
        product_barcode=product_barcode,
    )
    db.add(post)
    await db.flush()
    return {"id": post.id, "message": "Post created"}


@router.post("/posts/{post_id}/vote")
async def vote_post(
    post_id: str,
    body: VoteRequest,
    _: None = Depends(vote_limiter),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(select(CommunityPost).where(CommunityPost.id == post_id))
    post = result.scalar_one_or_none()
    if not post:
        raise HTTPException(status_code=404, detail="Post not found")
    if body.direction == "up":
        post.upvotes += 1
    elif body.direction == "down":
        post.downvotes += 1
    else:
        raise HTTPException(status_code=422, detail="direction must be 'up' or 'down'")
    await db.flush()
    return {"upvotes": post.upvotes, "downvotes": post.downvotes}


@router.get("/recipes")
async def list_recipes(db: AsyncSession = Depends(get_db), current_user: User = Depends(get_current_user)):
    result = await db.execute(
        select(CommunityPost)
        .where(CommunityPost.category == "recipe")
        .order_by(desc(CommunityPost.upvotes))
        .limit(20)
    )
    return {"posts": [{"id": p.id, "title": p.title, "upvotes": p.upvotes} for p in result.scalars().all()]}


@router.get("/symptoms")
async def list_symptom_threads(db: AsyncSession = Depends(get_db), current_user: User = Depends(get_current_user)):
    result = await db.execute(
        select(CommunityPost)
        .where(CommunityPost.category == "symptom")
        .order_by(desc(CommunityPost.created_at))
        .limit(20)
    )
    return {"posts": [{"id": p.id, "title": p.title, "created_at": p.created_at.isoformat()} for p in result.scalars().all()]}

File: backend/src/routes/community.js
import { Router } from 'express';
import { z } from 'zod';
import { requireAuth } from '../middleware/auth.js';
import { rateLimit, validateBarcode } from '../security.js';
import { store } from '../store.js';
import { HttpError, isoNow, newId, validateBody, validateQuery } from '../utils.js';

const VALID_CATEGORIES = ['general', 'recipe', 'symptom'];
const router = Router();
const writeLimiter = rateLimit({ limit: 20, windowSeconds: 60 });
const voteLimiter = rateLimit({ limit: 60, windowSeconds: 60 });
const postSchema = z.object({
  category: z.string().default('general'), title: z.string().min(5).max(160), body: z.string().min(1).max(5000),
  tags: z.array(z.string().nullable()).max(10).nullable().optional().default([]),
  product_barcode: z.string().max(50).nullable().optional().default(null),
});
const voteSchema = z.object({ direction: z.enum(['up', 'down']) });
const listSchema = z.object({ category: z.string().optional(), page: z.coerce.number().int().min(1).max(1000).default(1),
  per_page: z.coerce.number().int().min(1).max(100).default(15) });
router.use(requireAuth);

router.get('/posts', validateQuery(listSchema), (req, res) => {
  const { category, page, per_page: perPage } = req.validatedQuery;
  const filter = VALID_CATEGORIES.includes(category) ? category : null;
  const posts = store.getPosts(filter).slice((page - 1) * perPage, page * perPage);
  res.json({ posts: posts.map((post) => ({ id: post.id, author: post.author_name || 'Anonymous', category: post.category,
    title: post.title, body: post.body.slice(0, 200) + (post.body.length > 200 ? '...' : ''), tags: post.tags,
    upvotes: post.upvotes, downvotes: post.downvotes, net_votes: post.upvotes - post.downvotes,
    product_barcode: post.product_barcode, created_at: post.created_at })) });
});

router.post('/posts', writeLimiter, validateBody(postSchema), (req, res) => {
  const body = req.validatedBody;
  if (!VALID_CATEGORIES.includes(body.category)) throw new HttpError(422, "category must be one of ['general', 'recipe', 'symptom']");
  const now = isoNow();
  const post = { id: newId(), user_id: req.currentUser.id,
    author_name: req.currentUser.full_name || req.currentUser.email.split('@')[0], category: body.category,
    title: body.title.trim(), body: body.body.trim(), tags: (body.tags ?? []).map((tag) => (tag ?? '').trim().slice(0, 40)).filter(Boolean).slice(0, 10),
    product_barcode: body.product_barcode ? validateBarcode(body.product_barcode) : null,
    upvotes: 0, downvotes: 0, created_at: now, updated_at: now };
  store.savePost(post);
  res.status(201).json({ id: post.id, message: 'Post created' });
});

router.post('/posts/:postId/vote', voteLimiter, validateBody(voteSchema), (req, res) => {
  const post = store.getPostById(req.params.postId);
  if (!post) throw new HttpError(404, 'Post not found');
  if (req.validatedBody.direction === 'up') post.upvotes += 1;
  else post.downvotes += 1;
  res.json({ upvotes: post.upvotes, downvotes: post.downvotes });
});

router.get('/recipes', (_req, res) => {
  const recipes = store.getPosts('recipe').sort((a, b) => b.upvotes - a.upvotes).slice(0, 20);
  res.json({ posts: recipes.map((post) => ({ id: post.id, title: post.title, upvotes: post.upvotes })) });
});

router.get('/symptoms', (_req, res) => {
  const symptoms = store.getPosts('symptom').slice(0, 20);
  res.json({ posts: symptoms.map((post) => ({ id: post.id, title: post.title, created_at: post.created_at })) });
});

export default router;
